### ISSUE TITLE
write data to hdfs  when append=True, aways failed

### ISSUE THREAD

**Issue by banjin** [mtth/hdfs#92]
with client.write('/hadoop/stss.log',overwrite=False,append=True) as writer:
        #f.write(data_str)
        dump(data_str,writer)

Traceback (most recent call last):
  File "test-hdfs.py", line 14, in <module>
    main()
  File "test-hdfs.py", line 11, in main
    dump(data_str,writer)
  File "/usr/local/lib/python2.7/dist-packages/hdfs/util.py", line 100, in __exit__
    raise self._err # pylint: disable=raising-bad-type
hdfs.util.HdfsError: Failed to close file /hadoop/stss.log. Lease recovery is in progress. Try again later.
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.recoverLeaseInternal(FSNamesystem.java:3211)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInternal(FSNamesystem.java:3002)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInt(FSNamesystem.java:3287)
        at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFile(FSNamesystem.java:3251)
        at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.append(NameNodeRpcServer.java:618)
        at org.apache.hadoop.hdfs.server.namenode.AuthorizationProviderProxyClientProtocol.append(AuthorizationProviderProxyClientProtocol.java:126)
        at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.append(ClientNamenodeProtocolServerSideTranslatorPB.java:422)
        at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:617)
        at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1073)
        at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2217)
        at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2213)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:415)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1917)
        at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2211)

**Comment by mtth** [comment]
I'm not familiar with this error but it sounds like there are concurrent appends to the same file which isn't always supported. I don't think there's much we can do on the client side but you can apparently [tweak your cluster's configuration](https://community.hortonworks.com/articles/16144/write-or-append-failures-in-very-small-clusters-un.html) to support it.

**Comment by Tanmoy248** [comment]
I am sorry to have not responded on this thread before. As correctly pointed out, the issue is not with HDFS, but how hdfs manages its write pipeline. I found out that it was a unstable cluster that had packet losses between nodes. The problem was that the write pipeline once it fails, retries to write more datanodes. If it fails to find more datanode in the cluster (as in my dev cluster), it does not release the lock. The property to change this behavior `dfs.client.block.write.replace-datanode-on-failure.enable=false`
Temporary fix: I have to `hdfs cp` to a new file. And delete the corrupt file. I also had to change `fs.client.block.write.replace-datanode-on-failure.policy=NEVER`. 
Permanent fix: the interconnect between machines needs to be stabilized. Increase more datanodes in the cluster.
